File: src/BlockTree/BlockIndex.cs
using Gma.DataStructures;
using System;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Linq;

namespace TheBlockTree
{
	// no verification of block signatures nor chaining is done by the index
	public sealed class BlockIndex
	{
		private readonly Dictionary<ReadOnlyMemory<Byte>, Block> blocksBySignature =
			new Dictionary<ReadOnlyMemory<Byte>, Block>(ReadOnlyMemoryEqualityComparer<Byte>.Instance);

		private readonly Dictionary<ReadOnlyMemory<Byte>, OrderedSet<Block>> childBlocksByParentSignature =
			new Dictionary<ReadOnlyMemory<Byte>, OrderedSet<Block>>(ReadOnlyMemoryEqualityComparer<Byte>.Instance);

		public Int32 Count => blocksBySignature.Count;

		public Boolean Contains(ReadOnlyMemory<Byte> signature) =>
			blocksBySignature.ContainsKey(signature);

		public Boolean TryGetBySignature(ReadOnlyMemory<Byte> signature, [NotNullWhen(true)] out Block? block) =>
			blocksBySignature.TryGetValue(signature, out block);

		public Boolean TryGetChildren(ReadOnlyMemory<Byte> signature, [NotNullWhen(true)] out IReadOnlyList<Block>? children)
		{
			if (childBlocksByParentSignature.TryGetValue(signature, out var values))
			{
				children = values.ToList();
				return true;
			}
			children = null;
			return false;
		}

		public void Add(Block block)
		{
			blocksBySignature.Add(block.Signature, block);
			if (childBlocksByParentSignature.TryAddValue(block.ParentSignature, block))
				return;
			blocksBySignature.Remove(block.Signature);
			throw new ArgumentException("An element with the same key already exists in the set");
		}

		public IReadOnlyList<Block> GetAllBlocks() =>
			blocksBySignature.Values.ToList();

		public IEnumerable<(Block Block, UInt32 Level)> TraverseDepthFirst(Block root) => root.TraverseDepthFirst(
			x => TryGetChildren(x.Signature, out var children) ? children : Array.Empty<Block>()
		);

		public IEnumerable<(Block Block, UInt32 Level)> TraverseBreadthFirst(Block root) => root.TraverseBreadthFirst(
			x => TryGetChildren(x.Signature, out var children) ? children : Array.Empty<Block>()
		);
	}
}

File: src/typescript/BlockTree/BlockIndex.ts
import { Block } from "./Block";
import { Convert } from "./Convert";
import { NodeLevel, Traverse } from "./Traverse";

export class BlockIndex {
	private readonly blocksBySignature = new Map<string, Block>();
	private readonly childBlocksByParentSignature = new Map<string, Map<string, Block>>();

	contains(signature: Uint8Array): boolean {
		const key = Convert.toBase64String(signature);
		return this.blocksBySignature.get(key) !== undefined;
	}

	tryGetBySignature(signature: Uint8Array): Block | undefined {
		const key = Convert.toBase64String(signature);
		return this.blocksBySignature.get(key);
	}

	tryGetChildren(signature: Uint8Array): readonly Block[] | undefined {
		const key = Convert.toBase64String(signature);
		const set = this.childBlocksByParentSignature.get(key);
		if (set === undefined)
			return undefined;
		return Array.from(set.values());
	}

	add(block: Block): void {
		const key = Convert.toBase64String(block.signature);
		if (this.blocksBySignature.has(key))
			throw "Block with this signature already added";
		this.blocksBySignature.set(key, block);

		const parentKey = Convert.toBase64String(block.parentSignature);
		const childBlocks = this.childBlocksByParentSignature.get(parentKey);
		if (childBlocks === undefined)
			this.childBlocksByParentSignature.set(parentKey, new Map<string, Block>([[key, block]]));
		else
		{
			if (childBlocks.has(key)) {
				this.blocksBySignature.delete(key);
				throw "Block with this parent signature already added";
			}
			childBlocks.set(key, block);
		}
	}

	getAllBlocks(): readonly Block[] {
		return Array.from(this.blocksBySignature.values());
	}

	private getChildren(node: Block): IterableIterator<Block> {
		const children = this.tryGetChildren(node.signature);
		if (children === undefined)
			return [].values();
		return children.values();
	}

	traverseDepthFirst(root: Block): IterableIterator<NodeLevel<Block>> {
		return Traverse.depthFirst(root, this.getChildren);
	}

	traverseBreadthFirst(root: Block): IterableIterator<NodeLevel<Block>> {
		return Traverse.breadthFirst(root, this.getChildren);
	}
}